server/src/apub/shared_inbox.rs

```diff
@@ -0,0 +1,241 @@
+use super::*;
+
+#[serde(untagged)]
+#[derive(Serialize, Deserialize, Debug)]
+pub enum SharedAcceptedObjects {
+  Create(Create),
+  Update(Update),
+}
+
+/// Handler for all incoming activities to user inboxes.
+pub async fn shared_inbox(
+  request: HttpRequest,
+  input: web::Json<SharedAcceptedObjects>,
+  db: DbPoolParam,
+  chat_server: ChatServerParam,
+) -> Result<HttpResponse, Error> {
+  // TODO: would be nice if we could do the signature check here, but we cant access the actor property
+  let input = input.into_inner();
+  let conn = &db.get().unwrap();
+
+  let json = serde_json::to_string(&input)?;
+  debug!("Shared inbox received activity: {:?}", &json);
+
+  match input {
+    SharedAcceptedObjects::Create(c) => handle_create(&c, &request, &conn, chat_server),
+    SharedAcceptedObjects::Update(u) => handle_update(&u, &request, &conn, chat_server),
+  }
+}
```

Comment:
@asonix Do you know of anyway that I could switch on both the `kind`, and the action?

I have a lot of duplicated code in this file, because both `handle_create` and `handle_update` have to check the base_box kind (IE see if the incoming thing is either a post or a comment).

Reply:
if you had your own custom activity type like
```
#[derive(serde::Deserialize)]
struct LemmyActivity {
  #[serde(rename = "type")]
  kind: String,

  // enforce that valid lemmy activities have `object` fields
  // this may be too strict, but you could make an enum over Url | BaseBox for this
  object: BaseBox,

  /// ...etc
}
```
Then you could do 

```
match (activity.kind, activity.base_box.kind()) {
  ("Create", Some("Page")) => /* do something */
  ("Create", Some("Note")) => /* do something else */
  ...
}
```

Reply:
you could also do something with your current enum

```rust
impl SharedAcceptedObjects {
  fn object(&self) -> Option<&BaseBox> {
    match self {
      SharedAcceptedObjects::Create(c) => c.create_props.get_object_base_box(),
      SharedAcceptedObjects::Update(u) => u.update_props.get_object_base_box(),
      ...
    }
  }
}
```

and then do
```rust
let object = activity.object().cloned()?;
match (activity, object.kind()) {
  (SharedAcceptedObjects::Create(c), Some("Note")) => /* whatever */
  ...
}
```

Reply:
Sweet thanks, that second solution will work great.